### PR TITLE
Server Side Rendering

### PR DESCRIPTION
**Pull request by rupertbates**
## Why are you doing this?

To enable us to use Server Side Rendering (SSR) with our pages

To add a new SSR page

* Add a new entry into the array in `ssrPages.js` describing your page:
```
export const pages = [
  {
    filename: 'showcase.html',
    html: render(showcase),
  },
  {
    filename: 'paper-subscription-landing.html',
    html: render(paper),
  },
];
```
`filename` can be any unique and descriptive name, `html` should be a React element suitable to pass to `ReactDOMServer.renderToString()`

* Change the controller action that renders your page to call `assets.getSsrCacheContentsAsHtml()` with the id of the main page div and the filename declared in the previous step:
```
Ok(views.html.main(
      title = "Support the Guardian",
      mainElement = assets.getSsrCacheContentsAsHtml("showcase-landing-page","showcase.html"),
      ...
)
```

[**Trello Card**](https://trello.com/c/W4P97Vbs/2318-spike-how-do-we-make-our-site-faster)

### CODE REVIEW

File: support-frontend/webpack.ssr.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+const webpack = require('webpack');
+const merge = require('webpack-merge');
+const common = require('./webpack.common.js');
+
+module.exports = merge(common('[name].css', '[name].js', false,
+  [{
+    test: /(showcase.jsx|paperSubscriptionLandingPage.jsx)/,
```
Comment (walaura) [inline support-frontend/webpack.ssr.js]: this could be `pages/*.jsx` in the future for simplicity

File: support-frontend/webpack.common.js [hunk 2]
```diff
@@ -169,7 +169,7 @@ module.exports = (cssFilename, outputFilename, minimizeCss) => ({
           },
           ...cssLoaders,
         ],
-      },
-    ],
+      }
+    ].concat(additionalModuleRules || []),
```
Comment (walaura) [inline support-frontend/webpack.common.js]: nit

```suggestion
          ...additionalModuleRules,
        ],
```

File: support-frontend/scripts/ssr.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+//------------- Showcase page --------------
+
+const { getHtml: getShowcase } = require('../public/compiled-assets/javascripts/showcasePage');
+
+writePage('uk/support', getShowcase());
+
+//------------- Paper landing page --------------
+
+window.guardian.productPrices = {
```
Comment (walaura) [inline support-frontend/scripts/ssr.js]: not a huge fan of this for launch, they are nullable in the latest master so they can be empty and I think that's cleaner than hardcoding them

File: support-frontend/scripts/ssr.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+writePage('uk/subscribe/paper', getPaper());
+
+
+console.log('Done');
+process.exit();
+
+
+
+
```
Comment (walaura) [inline support-frontend/scripts/ssr.js]: this file could probably benefit from some cleanup but the overall principle looks solid!

File: support-frontend/assets/stylesheets/skeleton/html.scss [hunk 5]
```diff
@@ -77,3 +77,13 @@ Assistive technologies guidance (non-visible)
 .accessibility-hint {
   @include accessibility-hint;
 }
+
+
+/*
+Disable forms pre hydration
+*/
+[data-not-hydrated] {
```
Comment (walaura) [inline support-frontend/assets/stylesheets/skeleton/html.scss]: we should document this in the wiki

File: support-frontend/assets/pages/paper-subscription-landing/paperSubscriptionLandingPage.jsx [hunk 6]
```diff
@@ -29,6 +29,7 @@ import type { PaperFulfilmentOptions } from 'helpers/productPrice/fulfilmentOpti
 import { Collection, HomeDelivery } from 'helpers/productPrice/fulfilmentOptions';
 import { paperHasDeliveryEnabled } from 'helpers/subscriptions';
 import ConsentBanner from 'components/consentBanner/consentBanner';
+import ReactDOMServer from 'react-dom/server';
```
Comment (walaura) [inline support-frontend/assets/pages/paper-subscription-landing/paperSubscriptionLandingPage.jsx]: this is gonna bump up the filesize for the browser, could we instead return `content` from this file and import `ReactDOMServer`  from the `ssr.js` file?

File: .gitignore [hunk 7]
```diff
@@ -26,6 +26,8 @@ assets/javascripts/lib/bower-components/
 
 support-frontend/conf/keys.conf
 support-frontend/conf/assets.map
+support-frontend/conf/ssr-cache/
```
[the remark this answers is not in the document]
Comment (rupertbates) [reply .gitignore]: conf is the standard scala folder setup so we shouldn't change it

[the remark this answers is not in the document]
Comment (walaura) [reply .gitignore]: okay! didnt knew it was an idiom. let's keep it

**Pull request merged**